### ISSUE TITLE
Is it possible to download "Fast pass" and "locked" chapters?

### ISSUE THREAD

**Issue by Seriwes1** [Zehina/Webtoon-Downloader#47]
Is it possible to download "Fast pass" and "locked" chapters?

**Comment by Zehina** [comment]
Not sure. Definitely not something publicly available. Leaving open if anyone wants to investigate it in the future

**Comment by webtoontrade** [comment]
Yes, it is possible, but I don't think I will post how here. To download the chapters you need to reverse engineer the app. If webtoon notice a lot of people downloading the fast pass they might change the api.

**Comment by TadyHUN** [comment]
I'd like to know how bc I have a sries I want to dl but 54 chapter is fast pass/ad with only 6 bveing public.

**Comment by webtoontrade** [comment]
> [TadyHUN](/TadyHUN)

I have already explained above that you need to have paid for the chapter or watched an ad to download it. If you are asking if it is possible to download it without paying then that requires hacking webtoons servers and is illegal.

**Comment by 93Pd9s8Jt** [comment]
> Yes, it is possible, but I don't think I will post how here. To download the chapters you need to reverse engineer the app. If webtoon notices a lot of people downloading the fast pass they might change the api. You also need to be logged in and unlocked the chapter. As long as you unlocked the chapter it is possible to download fast pass and daily pass.

Do you know how? I've been looking at this recently. I don't think it should be too hard - I've already cracked the login, but I haven't yet managed to work out the authentication for https://global.apis.naver.com/lineWebtoon/webtoon/buyProduct.json, which is used to "buy" using a daily pass ticket. It then does GET /lineWebtoon/webtoon/episodeInfoWithLogin.json to fetch the images.  Fetching the images may also require some kind of authentication, I haven't checked.

**Comment by 93Pd9s8Jt** [comment]
Okay, from further testing, Webtoon's security is abysmal. For obvious reasons I won't be publically sharing what I've found, but I'm happy to share privately.

**Comment by TadyHUN** [comment]
#93Pd9s8Jt  how may we contact you?

**Comment by pustisobaka** [comment]
> Okay, from further testing, Webtoon's security is abysmal. For obvious reasons I won't be publically sharing what I've found, but I'm happy to share privately.

i'd also want to know

**Comment by pustisobaka** [comment]
@93Pd9s8Jt same question, how can i contact you?

**Comment by 93Pd9s8Jt** [comment]
@TadyHUN @pustisobaka you can email me at nshnfs@gmail.com. Apologies for not noticing your messages sooner.

**Comment by ngoomie** [comment]
I don't know much about how Webtoon works, but for downloading stuff you already have access to, maybe there could be a way to feed your Webtoon cookie into the program to let it download said chapters? Like what Deezloader does with downloading content off Deezer

**Comment by 93Pd9s8Jt** [comment]
@ngoomie yes, this is entirely possible. The relevant cookie is "NEO_SES".

**Comment by 93Pd9s8Jt** [comment]
I've made my code public at https://github.com/93Pd9s8Jt/webtoon-api. I do not currently intend to work on it, although that may change in the future.

**Comment by RyszardGmail** [comment]
> I've made my code public at https://github.com/93Pd9s8Jt/webtoon-api. I do not currently intend to work on it, although that may change in the future.

Your code gives "exception during login". Email account. Email account is correct. As I can expect it should work on similar level as android app, not PC website. And it should show bought episodes as well as unlocked ones.

About fast pass and locked chapters - I didn't test it much but from my short tests when Webtoons paywalls serie (there are many series which are free for now but it is mostly temporary status and Webtoons takes such series to paywall model) the links to pages are still valid but the links to chapters are removed. Links to page images are generated in the way there is no possible to recreate it without chapter website. So without chapter website we can't get links to images.
I think the app is using the same hosting as website. I tried to look into files generated by app but wasn't able to get any valid links to page images, found only thumb images links. It may be they are hashed but also they can be no cached at all.
I plan to modify some webtoon downloader, not necessary this one, and make it to no download but create text file with all links to images. But have too much other work now to do this in nearest future.
Chapters downloaded in native app have two flaws - they are lower quality (70 or 80%, 90% website) and they are intentionally damaged. Damages are reversible but need custom fix. And of course fast pass and unlocked chapters have no download function in the app.

You can also try to grab images from android cache but there are issues - names are hashed with something looking like sha256 algorithm and when app is taking image to cache it is not in the proper date order. So it is hard manual work to fix it. The plus is, cached images are the same quality as website version, and they are not damaged.

**Comment by 93Pd9s8Jt** [comment]
@RyszardGmail  Works just fine for me. The app does use the same domain as the website for free images, and a different one for paid ones. It's possible to get the original quality images using the app api. What do you mean the images are damaged?

**Comment by RyszardGmail** [comment]
@93Pd9s8Jt Yes, it works. Definition of encrypt in main.py misses self argument. Maybe it works with other python version, but not with 3.12. Maybe with UV, don't know. I launched it directly under python, without uv.

About damaged images I meant images downloaded with original app. Free series have download function, You can download for 30 days. Images are saved in phone /data folder and can be fetched if phone is rooted. But they are intentionally damaged and require fix.

I have some doubts - Your app shows ad pass for any series I have checked, for every chapters, but when I checked them on original apps, the ads were for everything except last chapters of series. Is it limitation of original app, where in reality api allows ad pass for all chapters?
Anyway, great piece of work.

**Comment by 93Pd9s8Jt** [comment]
@RyszardGmail oops, I cleaned up the code when uploading it to github and didn't test it. Fixed now.
The images are undamaged when fetched from the api.
Can you give some examples for ad pass being wrong? I didn't write the code for the tkinter app; I can't verify that it uses the api correctly, but it's been correct for all the things I've tested.

**Comment by RyszardGmail** [comment]
@93Pd9s8Jt Can I send private message on github?

I will put additional infos (if i get anything new) in your github project. I have created issue, it is now closed but I will continue to write this or that there. As I understand You get notifications with new comments, even if thread is closed.

**Comment by 93Pd9s8Jt** [comment]
Yes, you can continue commenting in the closed issue there.
